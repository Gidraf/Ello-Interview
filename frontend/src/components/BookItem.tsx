import React from 'react';
import { Book } from '../types';
import { Card, CardContent, Typography, Button, CardMedia, CardActions } from '@mui/material';




interface BookItemProps {
  book: Book;
  addBookToReadingList: (book: Book) => void;
  readList: any;
  removeBookFromReadingList: (title: string) => void
}

const BookItem: React.FC<BookItemProps> = ({ book, addBookToReadingList, readList, removeBookFromReadingList }) => {
    console.log(book.title)
    console.log(readList)
  return (
    <Card style={{ marginBottom: '16px' }}>
       <CardMedia
        sx={{ height: 140 }}
        image={require(`../${book.coverPhotoURL}`)}
        title="green iguana"
      />
      <CardContent>
        <Typography variant="h5">{book.title}</Typography>
        <Typography variant="subtitle1">{book.author}</Typography>
        { !readList[book.title] ? <Button variant="contained" color="primary" onClick={() => addBookToReadingList(book)}>
           Add to Reading List
         </Button> : <Button variant="contained" color="secondary" onClick={() => removeBookFromReadingList(book.title)}>
Remove
</Button>
         
         }
      </CardContent>
    </Card>
  );
};

export default BookItem;
