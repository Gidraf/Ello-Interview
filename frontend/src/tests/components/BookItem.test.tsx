import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import BookItem from '../../components/BookItem';


const mockBook = {
  title: 'Test Book',
  author: 'Test Author',
  coverPhotoURL: 'assets/image10.webp',
  readingLevel:"D"
};

const addBookToReadingList = jest.fn();
const removeBookFromReadingList = jest.fn();

describe('BookItem Component', () => {
  it('should match the snapshot when book is not in the reading list', () => {
    const { asFragment } = render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{}}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match the snapshot when book is in the reading list', () => {
    const { asFragment } = render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{ 'Test Book': true }}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display book title and author', () => {
    render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{}}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );

    expect(screen.getByText('Test Book')).toBeInTheDocument();
    expect(screen.getByText('Test Author')).toBeInTheDocument();
  });

  it('should display add button when book is not in the reading list', () => {
    render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{}}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );

    const addButton = screen.getByText('Add to Reading List');
    expect(addButton).toBeInTheDocument();
    fireEvent.click(addButton);
    expect(addBookToReadingList).toHaveBeenCalledWith(mockBook);
  });

  it('should display remove button when book is in the reading list', () => {
    render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{ 'Test Book': true }}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );

    const removeButton = screen.getByText('Remove');
    expect(removeButton).toBeInTheDocument();
    fireEvent.click(removeButton);
    expect(removeBookFromReadingList).toHaveBeenCalledWith('Test Book');
  });

  it('should display book cover image', () => {
    render(
      <BookItem
        book={mockBook}
        addBookToReadingList={addBookToReadingList}
        readList={{}}
        removeBookFromReadingList={removeBookFromReadingList}
      />
    );

    const coverImage = screen.getByTitle('green iguana');
    expect(coverImage).not.toBe(null);
  });
});
