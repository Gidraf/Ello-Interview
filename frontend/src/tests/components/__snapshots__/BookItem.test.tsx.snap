// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem Component should match the snapshot when book is in the reading list 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="margin-bottom: 16px;"
  >
    <div
      class="MuiCardMedia-root css-1999db7-MuiCardMedia-root"
      role="img"
      style="background-image: url(image10.webp);"
      title="green iguana"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Test Book
      </h5>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
      >
        Test Author
      </h6>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Remove
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BookItem Component should match the snapshot when book is not in the reading list 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="margin-bottom: 16px;"
  >
    <div
      class="MuiCardMedia-root css-1999db7-MuiCardMedia-root"
      role="img"
      style="background-image: url(image10.webp);"
      title="green iguana"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Test Book
      </h5>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
      >
        Test Author
      </h6>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add to Reading List
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
